packages/rest-client/src/main/RestClient.ts

```diff
@@ -0,0 +1,184 @@
+import { HttpAgent, HttpsAgent } from 'agentkeepalive'
+import superagent, { ResponseError } from 'superagent'
+import { Readable } from 'stream'
+import type Logger from 'bunyan'
+import sanitiseError from './helpers/sanitiseError'
+import { ApiConfig } from './types/ApiConfig'
+import { AuthOptions, Token } from './types/AuthOptions'
+import { Request, RequestWithBody, StreamRequest } from './types/Request'
+
+export default abstract class RestClient {
+  private readonly agent: HttpAgent
+
+  protected constructor(
+    private readonly name: string,
+    private readonly config: ApiConfig,
+    private readonly logger: Logger | Console,
+    private readonly getSystemToken: (username?: string) => Promise<string>,
+  ) {
+    this.agent = config.url.startsWith('https') ? new HttpsAgent(config.agent) : new HttpAgent(config.agent)
+  }
+
+  private apiUrl() {
+    return this.config.url
+  }
+
+  private timeoutConfig() {
+    return this.config.timeout
+  }
+
+  private handleRetry(retry: boolean = true) {
+    return (err: Error) => {
+      if (!retry) {
+        return false
+      }
+
+      if (err) {
+        this.logger.info(`Retry handler found API error with ${err.name} - ${err.message}`)
+      }
+
+      return undefined
+    }
+  }
+
+  async get<Response = unknown>(
+    { path, query = {}, headers = {}, responseType = '', raw = false }: Request,
+    authOptions: AuthOptions,
+  ): Promise<Response> {
+    this.logger.info(`${this.name} GET: ${path}`)
+
+    const token = await this.resolveToken(authOptions)
+
+    try {
+      const result = await superagent
+        .get(`${this.apiUrl()}${path}`)
+        .query(query)
+        .agent(this.agent)
+        .retry(2, this.handleRetry())
+        .auth(token, { type: 'bearer' })
+        .set(headers)
+        .responseType(responseType)
+        .timeout(this.timeoutConfig())
```

Comment:
Coercing 404s to null is a pretty common customisation.. we probably need to support that at some point but not necessarily now!

We could do a beta and try it out in a few projects and see if any common customisations occur for other things.. number of retries might be one